### PR TITLE
Add some unittest on get_buffer

### LINKED ISSUE #886: Content offset is ignored on macOS using LibZim 9.2.0
Under macOS Kiwix app, using **LibZIM 9.2.0**

We have the [following code](https://github.com/kiwix/kiwix-apple/blob/main/Model/ZimFileService/ZimFileService.mm#L199) to read the content from the zim file using offset and size:
```
blob = item.getData(start, fmin(item.getSize() - start, end - start + 1));
```
Which if I got this right has the corresponding C++ function here:
https://github.com/openzim/libzim/blob/main/src/item.cpp#L50C1-L50C61

My aim is to read the data in chunks, but it seems there's a bug with the offset.
Reading the entire data from `offset=0` to `item.getSize()` works as expected.
Starting with `offset=0` and increasing the `size` also works as expected.

Changing the offset, does not work, it seems to be ignored, I am keep getting back the same data.
I did tried it with `text/html`, `webp image` and what I would like most to be working is `webm` video content.

Here are some debug prints from `webm` chunks:
```
"getURLContent: /videos/9987/video.webm 2076400 - 2076464: 1a45dfa39f4286810142f7810142f2810442f381084282847765626d4287810242858102185380670100000000f1a98a114d9b74bb4dbb8b53ab841549a96653"

"getURLContent: /videos/9987/video.webm 2076600 - 2076664: 1a45dfa39f4286810142f7810142f2810442f381084282847765626d4287810242858102185380670100000000f1a98a114d9b74bb4dbb8b53ab841549a96653"
```

It is the same on HTML as well read in chunks sized 100:
```
"Optional(\"text/html\") -> Optional(7369)"
"mime: text/html"

mType text/html from 0 to 99, size: 100
"getURLContent: /A/App/IntroPage 0 - 99: Optional(\"<!DOCTYPE html>\\n<html class=\\\"client-js\\\"><head>\\n  <meta charset=\\\"UTF-8\\\">\\n  <title>App/IntroPage</titl\")"
"got content @ isMain: false from: 0 - 99 with data: 100"

mType text/html from 100 to 199, size: 100
"getURLContent: /A/App/IntroPage 100 - 199: Optional(\"<!DOCTYPE html>\\n<html class=\\\"client-js\\\"><head>\\n  <meta charset=\\\"UTF-8\\\">\\n  <title>App/IntroPage</titl\")"
"got content @ isMain: false from: 100 - 199 with data: 100"

mType text/html from 200 to 299, size: 100
"getURLContent: /A/App/IntroPage 200 - 299: Optional(\"<!DOCTYPE html>\\n<html class=\\\"client-js\\\"><head>\\n  <meta charset=\\\"UTF-8\\\">\\n  <title>App/IntroPage</titl\")"
"got content @ isMain: false from: 200 - 299 with data: 100"
```

Note: I did tested it with a couple of unrelated ZIM files, and the offset was still ignored. 
I even hard coded the offset values, and that still resulted in the same data returned.

### PR DESCRIPTION
**Pull request by mgautierfr**
This Pr was initially to fix #886. But it appears it was not a bug on our side.

However, it is better testing and it reveals some bug on Windows side.

### CODE REVIEW

**Comment by kelson42** [comment]
Where is rhe codecov, it seems it has vanished!

**Comment by rgaudin** [comment]
I think it doesn't run on drafts

File: src/fs_unix.cpp [hunk 1]
```diff
@@ -47,8 +47,11 @@ zsize_t FD::readAt(char* dest, zsize_t size, offset_t offset) const
   errno = 0;
   while (size_to_read > 0) {
     auto size_read = PREAD(m_fd, dest, size_to_read, current_offset);
+    if (size_read == 0) {
```
Comment (veloman-yunkan) [inline src/fs_unix.cpp:50]: This is comment on the commit message:

- excepts → expects
- is → if

**Review by veloman-yunkan — CHANGES_REQUESTED** [review]

File: src/fs_windows.cpp [hunk 2]
```diff
@@ -73,22 +73,43 @@ zsize_t FD::readAt(char* dest, zsize_t size, offset_t offset) const
...
+      errorMsg = "Cannot read past the end of the file";
+	  goto err;
+    }
+
+    size_to_read -= size_read;
+    full_size_read += size_read;
+	dest += size_read;
   }
-  if (size_read != size.v) {
+  if (full_size_read != size.v) {
```
[the remark this answers is not in the document]
Comment (mgautierfr) [reply src/fs_windows.cpp:104]: Indeed. `full_size_read` removed.

File: src/fs_windows.cpp [hunk 3]
```diff
@@ -73,22 +73,43 @@ zsize_t FD::readAt(char* dest, zsize_t size, offset_t offset) const
...
+  while (size_to_read > 0) {
+	// Read by batch < 4GiB
+	// Lets use a batch of 4MiB
+	auto batch_to_read = std::min(size_to_read, (size_type)4*1024*1024);
+    if (!ReadFile(mp_impl->m_handle, dest, batch_to_read, &size_read, NULL)) {
+	  errorMsg = "Read fail";
+      goto err;
+    }
+
+	if (size_read == 0) {
```
[the remark this answers is not in the document]
Comment (mgautierfr) [reply src/fs_windows.cpp]: From `ReadFile` function, it is technically possible than `size_read != batch_to_read`. If we reach the end of file or (at least on linux, don't known on windows) if we have been interrupted by a signal.

The eof should not happen as we have check validity of the read before.

But I prefer to keep it. If it appear it is really something wrong (unexpected eof), we would have a step more in the loop before throwing a error (and mostly stopping the program).

File: src/fs_windows.cpp [hunk 4]
```diff
@@ -73,22 +73,43 @@ zsize_t FD::readAt(char* dest, zsize_t size, offset_t offset) const
...
     goto err;
   }
 
   DWORD size_read;
-  if (!ReadFile(mp_impl->m_handle, dest, size.v, &size_read, NULL)) {
-    goto err;
+  while (size_to_read > 0) {
+	// Read by batch < 4GiB
+	// Lets use a batch of 4MiB
+	auto batch_to_read = std::min(size_to_read, (size_type)4*1024*1024);
```
[the remark this answers is not in the document]
Comment (mgautierfr) [reply src/fs_windows.cpp]: Indeed. Done.

**Review by veloman-yunkan — COMMENTED** [review]
Please rebase and fix up. I am almost certain that then it will LGTM.

File: src/reader.h [hunk 5]
```diff
@@ -31,13 +32,23 @@
...
+    void read(char* dest, offset_t offset, zsize_t size) const {
+      if (can_read(offset, size)) {
+        if (size) {
+          // Do the actuall read only if we have a size to read
+          readImpl(dest, offset, size);
+        }
+        return;
+      }
+      throw std::runtime_error("Cannot read after the end of the reader");
+    }
```
Comment (veloman-yunkan) [inline src/reader.h:50]: It would be nice if the commit message is updated

File: src/buffer_reader.h [hunk 6]
```diff
@@ -30,11 +30,11 @@ class LIBZIM_PRIVATE_API BufferReader : public Reader {
       : source(source) {}
     virtual ~BufferReader() {};
 
-    zsize_t size() const;
-    offset_t offset() const;
+    zsize_t size() const override;
+    offset_t offset() const override;
```
Comment (veloman-yunkan) [inline src/buffer_reader.h:34]: This commit better be moved right after the one introducing `Reader::readImpl()`

**Review by veloman-yunkan — COMMENTED** [review]
Looks good to me but not to my more picky self :)

Please merge after addressing the two concerns with the commit history

**Comment by mgautierfr** [comment]
> Looks good to me but not to my more picky self :)

Done. Hope picky @veloman-yunkan is happy with the change.

> Please merge after addressing the two concerns with the commit history

I need a explicit approval to merge. Please approve (and merge directly while you have the button under your mouse cursor)

**Review by kelson42 — APPROVED** [review]
LGTM ;)

**Pull request merged**